Fix inconsistent behavior for to_unixtime with timestamp

martint opened the pull request:
Fixes https://github.com/trinodb/trino/issues/7450

martint committed:
Fix inconsistent behavior for to_unixtime with timestamp

After the timestamp semantics fix, the behavior of the variant of the
function that takes timestamp(p) became inconsistent with the
timestamp(p) w/ time zone variant. Unix time represent a point in time
(vs a wall clock time), so, conceptually, a timestamp(p) first needs to
be converted to a timestamp(p) with time zone before the transformation
is applied.

Currently, such timestamps are interpreted as being UTC, which is confusing
and counter-intuitive, especially in comparison with an invocation that uses
an explicit cast:

    SELECT
      to_unixtime(localtimestamp),
      to_unixtime(cast(localtimestamp as timestamp with time zone))

          _col0       |      _col1
    ------------------+------------------
     1.617009524129E9 | 1.617034724129E9

This fixes the semantics of the function by relying on the standard coercion
rules to convert from timestamp(p) -> timestamp(p) w/ time zone

findepi commented on core/trino-main/src/test/java/io/trino/operator/scalar/timestamp/TestTimestamp.java:
```diff
@@ -63,19 +63,19 @@ public void teardown()
+        assertThat(assertions.expression("to_unixtime(TIMESTAMP '2020-05-10 12:34:56.12')")).matches("1589067296.12e0");
+        assertThat(assertions.expression("to_unixtime(TIMESTAMP '2020-05-10 12:34:56.123')")).matches("1589067296.123e0");
+        assertThat(assertions.expression("to_unixtime(TIMESTAMP '2020-05-10 12:34:56.1234')")).matches("1589067296.1234e0");
+        assertThat(assertions.expression("to_unixtime(TIMESTAMP '2020-05-10 12:34:56.12345')")).matches("1589067296.1234498e0");
```
i wish `to_unixtime` returned exact numeric ...

findepi commented on core/trino-main/src/test/java/io/trino/operator/scalar/timestamp/TestTimestamp.java:
```diff
@@ -63,19 +63,19 @@ public void teardown()
-        assertThat(assertions.expression("to_unixtime(TIMESTAMP '2020-05-10 12:34:56.1234567890')")).matches("1589114096.1234567e0");
-        assertThat(assertions.expression("to_unixtime(TIMESTAMP '2020-05-10 12:34:56.12345678901')")).matches("1589114096.1234567e0");
-        assertThat(assertions.expression("to_unixtime(TIMESTAMP '2020-05-10 12:34:56.123456789012')")).matches("1589114096.1234567e0");
+        assertThat(assertions.expression("to_unixtime(TIMESTAMP '2020-05-10 12:34:56')")).matches("1589067296e0");
```
Maybe add a comment we're testing to_unixtime + coercion here, as it's unusual.

findepi commented:
It's important that we follow up with `from_unixtime` changes, otherwise things will be inconsistent and confusing. 
People will expect the functions to roundtrip....

martint committed:
Return timestamp with time zone from from_unixtime

Unix time represents a point in time, so it is natural for this
function to return a timestamp with time zone instead of a
plain timestamp.

martint commented:
I added a commit for `from_unixtime`. Now the conversions roundtrip properly:

```before
trino> select 
    localtimestamp, 
    from_unixtime(to_unixtime(localtimestamp)), 
    current_timestamp, 
    from_unixtime(to_unixtime(current_timestamp)) \G

-[ RECORD 1 ]--------------------------------------
_col0 | 2021-03-30 15:50:52.543
_col1 | 2021-03-30 15:50:52.543
_col2 | 2021-03-30 15:50:52.543 America/Los_Angeles
_col3 | 2021-03-30 22:50:52.543
```

after:
```
trino> select 
    localtimestamp, 
    from_unixtime(to_unixtime(localtimestamp)), 
    current_timestamp, 
    from_unixtime(to_unixtime(current_timestamp)) \G

-[ RECORD 1 ]--------------------------------------
_col0 | 2021-03-30 15:49:26.375
_col1 | 2021-03-30 15:49:26.375 America/Los_Angeles
_col2 | 2021-03-30 15:49:26.375 America/Los_Angeles
_col3 | 2021-03-30 15:49:26.375 America/Los_Angeles
```

GuoPhilipse commented:
@martint @findepi Great to meet this new fix, do we have the plan for next release?

findepi commented on core/trino-main/src/main/java/io/trino/operator/scalar/timestamp/ToUnixTime.java:
```diff
@@ -1,43 +0,0 @@
-@ScalarFunction("to_unixtime")
-public final class ToUnixTime
-{
-    private ToUnixTime() {}
```
Update datetime.rst as well

```
.. function:: to_unixtime(timestamp) -> double

    Returns ``timestamp`` as a UNIX timestamp.
```

martint replied:
`timestamp` in that context is the name of the argument, not a type

findepi replied:
isn't that a bit less informative than it could be?
(not to say 'misleading')

martint replied:
There's no way to indicate the type in the rst syntax we use, unfortunately. Also, the names of the arguments are not very consistent. We have stuff like this, for instance: `.. function:: date(x) -> date`

findepi commented on core/trino-main/src/main/java/io/trino/operator/scalar/DateTimeFunctions.java:
```diff
@@ -121,10 +119,10 @@ public static Slice currentTimeZone(ConnectorSession session)
+    public static long fromUnixTime(ConnectorSession session, @SqlType(StandardTypes.DOUBLE) double unixTime)
     {
-        return Math.round(unixTime * MICROSECONDS_PER_SECOND);
+        return packDateTimeWithZone(Math.round(unixTime * 1000), session.getTimeZoneKey());
```
Some points in time are non-representable in the session zone.
maybe add
```
// TODO (https://github.com/trinodb/trino/issues/5781) ....
```


here

(it doesn't make sense to fix this problem for this method, but not fix it elsewhere, hence TODO only)

(https://github.com/trinodb/trino/issues/5781)

findepi commented on core/trino-main/src/main/java/io/trino/operator/scalar/DateTimeFunctions.java:
```diff
@@ -155,9 +153,10 @@ public static long fromUnixTime(@SqlType(StandardTypes.DOUBLE) double unixTime,
+        @SqlType("timestamp(9) with time zone")
+        public static LongTimestampWithTimeZone fromLong(@LiteralParameter("s") long scale, ConnectorSession session, @SqlType("decimal(p, s)") Slice unixTimeNanos)
         {
+            // TODO
```
?

findepi commented on core/trino-main/src/main/java/io/trino/operator/scalar/DateTimeFunctions.java:
```diff
@@ -167,27 +166,27 @@ public static LongTimestamp fromLong(@LiteralParameter("s") long scale, @SqlType
                 picosOfSecond += PICOSECONDS_PER_SECOND;
             }
-            return DateTimes.longTimestamp(epochSeconds, picosOfSecond);
+            return DateTimes.longTimestampWithTimeZone(epochSeconds, picosOfSecond, session.getTimeZoneKey().getZoneId());
```
as above -- Some points in time are non-representable in the session zone.

findepi commented on testing/trino-product-tests/src/main/resources/sql-tests/testcases/horology_functions/checkHorologyFunctionsRegistered.result:
```diff
@@ -66,7 +66,6 @@
  quarter | bigint | timestamp(p) with time zone | scalar | true | Quarter of the year of the given timestamp |
  timezone_hour | bigint | timestamp(p) with time zone | scalar | true | Time zone hour of the given timestamp |
  timezone_minute | bigint | timestamp(p) with time zone | scalar | true | Time zone minute of the given timestamp |
- to_unixtime | double | timestamp(p) | scalar | true | |
```
Shows up in `Return timestamp with time zone from from_unixtime` but belongs to prev commit
